# Introduction to Data Preparation for NYC Taxi Trips

## Objective
This Jupyter Notebook is dedicated to the initial stage of our data analytics project—**Data Preparation**. The primary goal here is to prepare the vast New York City Taxi Trips dataset for the year 2021, ensuring it's clean, organized, and ready for in-depth analysis and modeling in subsequent stages. The dataset, which includes over 30 million records, requires meticulous handling to manage its volume and enhance its quality effectively.

## Background
The NYC Taxi Trips dataset is sourced from the NYC Open Data portal and offers a detailed snapshot of taxi activities across New York City. It records every taxi trip's core details, such as times of pickup and dropoff, trip distances, fares charged, and more. These records not only provide insights into the city’s mobility patterns but also serve as a basis for predictive modeling of fares and understanding factors influencing taxi trip dynamics.

## Scope of This Notebook
In this notebook, we will perform several key tasks to prepare the data for further analysis:
1. **Data Loading**: Load the data from four pre-processed partitions to manage the dataset's size efficiently.
2. **Initial Exploration**: Conduct a preliminary examination to understand the dataset's structure, missing values, anomalies, and data types.
3. **Data Cleaning**: Address missing or incorrect values, remove duplicates, and handle any outliers or erroneous entries.
4. **Feature Engineering**: Develop new features that are more informative for analysis and predictive modeling, such as calculating trip durations and categorizing times of day.
5. **Data Transformation**: Standardize and normalize data as necessary to prepare for machine learning algorithms that require standardized input.
6. **Data Reduction**: Reduce dimensionality where applicable to improve model performance and decrease computational requirements.

## Tools and Libraries
We will use Python as our main programming language, leveraging libraries such as Pandas for data manipulation, Numpy for numerical operations, and Matplotlib/Seaborn for visualization purposes. These tools are chosen for their efficiency and ease of use in handling large datasets like ours.

## Conclusion
By the end of this notebook, the dataset will be transformed into a clean, comprehensive format suitable for detailed exploratory data analysis and machine learning tasks in the following stages of this project. The meticulous preparation we perform here is crucial for ensuring the accuracy and reliability of our later analyses and predictions.


# 1. Data Loading 
In this section we import all necessary libraries and datasets for the project.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Set some options for displaying the data tables nicely
pd.set_option('display.max_columns', None)  # Show all columns of DataFrames
pd.set_option('display.width', 1000)        # Ensure the display is wide enough to view all DataFrame columns
pd.set_option('display.float_format', '{:,.2f}'.format)  # Format floats for easier reading

# Setting the style for seaborn plots
sns.set(style="whitegrid")


Pickle is a Python-specific binary serialization method used to save and load Python objects directly, preserving their data types and structure. In our NYC Taxi Trips data analytics project, we chose Pickle for its efficiency and ease of use, especially given the large volume of data involved. It enables fast loading and saving of complex Pandas DataFrames, significantly speeding up our workflow by avoiding repeated preprocessing. Although Pickle should be used cautiously due to security risks when dealing with untrusted data sources, it is ideal for our controlled environment where these concerns are mitigated.



In [2]:
import pandas as pd

# Define data types for data consistency
dtypes = {
    'VendorID': 'category',
    'tpep_pickup_datetime': 'str',
    'tpep_dropoff_datetime': 'str',
    'passenger_count': 'float64',
    'trip_distance': 'float64',
    'RatecodeID': 'category',
    'store_and_fwd_flag': 'category',
    'PULocationID': 'category',
    'DOLocationID': 'category',
    'payment_type': 'category',
    'fare_amount': 'float64',
    'extra': 'float64',
    'mta_tax': 'float64',
    'tip_amount': 'float64',
    'tolls_amount': 'float64',
    'improvement_surcharge': 'float64',
    'total_amount': 'float64',
    'congestion_surcharge': 'float64'
}

# Load data from CSV files only if Pickle files do not exist or when processing for the first time
try:
    df1 = pd.read_pickle("/Users/md/Desktop/python_project/df1.pkl")
    df2 = pd.read_pickle("/Users/md/Desktop/python_project/df2.pkl")
    df3 = pd.read_pickle("/Users/md/Desktop/python_project/df3.pkl")
    df4 = pd.read_pickle("/Users/md/Desktop/python_project/df4.pkl")
    print("Data loaded from Pickle files.")
except FileNotFoundError:
    print("Pickle files not found. Loading data from CSV files and saving as Pickle.")
    df1 = pd.read_csv("2021_TLC_0.csv", dtype=dtypes)
    df2 = pd.read_csv("2021_TLC_1.csv", dtype=dtypes)
    df3 = pd.read_csv("2021_TLC_2.csv", dtype=dtypes)
    df4 = pd.read_csv("2021_TLC_3.csv", dtype=dtypes)
    # Save DataFrames to Pickle for future use
    df1.to_pickle("/Users/md/Desktop/python_project/df1.pkl")
    df2.to_pickle("/Users/md/Desktop/python_project/df2.pkl")
    df3.to_pickle("/Users/md/Desktop/python_project/df3.pkl")
    df4.to_pickle("/Users/md/Desktop/python_project/df4.pkl")



Data loaded from Pickle files.


In [3]:
df1.head(5)

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance RatecodeID store_and_fwd_flag PULocationID DOLocationID payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  improvement_surcharge  total_amount  congestion_surcharge
0      1.0  2021-01-01 00:30:10   2021-01-01 00:36:12             1.00           2.10        1.0                  N        142.0         43.0          2.0         8.00   3.00     0.50        0.00          0.00                   0.30         11.80                  2.50
1      1.0  2021-01-01 00:51:20   2021-01-01 00:52:19             1.00           0.20        1.0                  N        238.0        151.0          2.0         3.00   0.50     0.50        0.00          0.00                   0.30          4.30                  0.00
2      1.0  2021-01-01 00:43:30   2021-01-01 01:11:06             1.00          14.70        1.0                  N        132.0        165.0          1.0        42.00   0.50     0.50        8.65          0.00                   0.30         51.95                  0.00
3      1.0  2021-01-01 00:15:48   2021-01-01 00:31:01             0.00          10.60        1.0                  N        138.0        132.0          1.0        29.00   0.50     0.50        6.05          0.00                   0.30         36.35                  0.00
4      2.0  2021-01-01 00:31:49   2021-01-01 00:48:21             1.00           4.94        1.0                  N         68.0         33.0          1.0        16.50   0.50     0.50        4.06          0.00                   0.30         24.36                  2.50

In the section above we loaded all 4 partitioned datasets for analysis defined data types based on the documentation for the dataset. We then loaded the Dataframes into pickle format for faster processing and for the purpose of tracking and maintaining consistency throughout rest of the notebook without the need to load large CSV files into df after every launch.

# 2. Initial Exploration

After loading the data, it's crucial to understand its structure, identify any missing values, spot potential anomalies, and review the data types. We start by exploring one of the DataFrames:



### Partition 1 Initial Exploration

In [4]:
# Display the first few rows of the first DataFrame
print(df1.head(3))

# Display summary information about DataFrame
print(df1.info())

# Basic statistical details
print(df1.describe())


  VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance RatecodeID store_and_fwd_flag PULocationID DOLocationID payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  improvement_surcharge  total_amount  congestion_surcharge
0      1.0  2021-01-01 00:30:10   2021-01-01 00:36:12             1.00           2.10        1.0                  N        142.0         43.0          2.0         8.00   3.00     0.50        0.00          0.00                   0.30         11.80                  2.50
1      1.0  2021-01-01 00:51:20   2021-01-01 00:52:19             1.00           0.20        1.0                  N        238.0        151.0          2.0         3.00   0.50     0.50        0.00          0.00                   0.30          4.30                  0.00
2      1.0  2021-01-01 00:43:30   2021-01-01 01:11:06             1.00          14.70        1.0                  N        132.0        165.0          1.0        42.00   0.50     0.50        8.

The output from your DataFrame `df1` provides a snapshot of the NYC Taxi Trips dataset for 2021, along with a detailed summary and structure of the data. Here's an analysis of the provided output:

### Data Structure and Content Overview:
- **Rows and Columns:** The DataFrame contains approximately 7,919,804 entries across 18 different attributes (columns), suggesting a substantial amount of taxi trip data.
- **Data Types:** 
  - **Categorical Data:** Columns like `VendorID`, `RatecodeID`, `store_and_fwd_flag`, `PULocationID`, `DOLocationID`, and `payment_type` are stored as categories, which is efficient for columns with a limited number of unique values.
  - **Datetime Data:** The pickup and dropoff times (`tpep_pickup_datetime` and `tpep_dropoff_datetime`) are appropriately stored as `datetime64[ns]`, facilitating time series analysis.
  - **Numerical Data:** Attributes like `passenger_count`, `trip_distance`, `fare_amount`, and other cost-related fields are stored as `float64`, suitable for calculations and statistical analysis.

### Statistical Summary:
- **Central Tendency and Distribution:**
  - **`passenger_count`** shows an average of 1.41 passengers per trip, with most trips carrying just 1 passenger.
  - **`trip_distance`** has an average of 5.02 miles, but the standard deviation and the maximum value suggest extreme outliers (max 280,567.84 miles).
  - **`fare_amount`** has an average of has an average of 12.36, but similar to trip distance,the standard deviation and maximum suggest significant outliers or possible data entry errors (max $398,466.38).
  
- **Anomalies/Outliers:** 
  - The presence of negative values in `fare_amount`, `extra`, `mta_tax`, `tip_amount`, `tolls_amount`, `improvement_surcharge`, `total_amount`, and `congestion_surcharge` might indicate refunds, data entry errors, or other anomalies.
  - Extremely high maximum values in `trip_distance`, `fare_amount`, `tip_amount`, and `total_amount` could be due to erroneous entries or rare, very long trips.

### Potential Data Quality Issues:
- **Outliers and Errors:** As noted, the presence of extreme values and negative amounts could affect the integrity of analyses and models if not addressed.
- **Missing Values:** Any NaNs or nulls in the dataset would need to be identified and handled appropriately during data cleaning.

### Suggested Next Steps:
1. **Data Cleaning:** Address outliers and errors, possibly by removing or capping values based on reasonable thresholds or business understanding.
2. **Handling Missing Values:** If any missing values are identified (which should be checked explicitly), determine the best strategy for handling them—either through imputation or removal, depending on the context.
3. **Feature Engineering:** Consider deriving new features, such as trip duration, hour of the day, day of the week, which might be useful for further analysis and predictive modeling.
4. **Deeper Exploration:** Further exploratory data analysis would be beneficial to understand relationships between variables, especially how categorical features relate to fare and distance.


In [5]:
# Count missing values in each column
missing_values1 = df1.isnull().sum()
print("Missing values in each column:\n", missing_values1)


Missing values in each column:
 VendorID                 452538
tpep_pickup_datetime          0
tpep_dropoff_datetime         0
passenger_count          452538
trip_distance                 0
RatecodeID               452538
store_and_fwd_flag       452538
PULocationID                  0
DOLocationID                  0
payment_type             452538
fare_amount                   0
extra                         0
mta_tax                       0
tip_amount                    0
tolls_amount                  0
improvement_surcharge         0
total_amount                  0
congestion_surcharge          0
dtype: int64




#### Analysis of Missing Values

Several columns in your DataFrame have the exact same number of missing values (452,538), suggesting a pattern or potential issue during data collection or extraction. The affected columns are:

- **`VendorID`**
- **`passenger_count`**
- **`RatecodeID`**
- **`store_and_fwd_flag`**
- **`payment_type`**

Given that none of the monetary or distance-related columns (`fare_amount`, `trip_distance`, `total_amount`, etc.) have missing values, the missing data is restricted to categorical and count data, which could impact analyses that depend on these dimensions, such as demographic studies or rate code analysis.

##### Suggested Handling Strategies

1. **Understanding the Context:**
   - **Determine the Cause:** Before deciding on a handling strategy, investigate why these values are missing. Is it a systematic error, data corruption, or are they missing at random? This might involve checking back with the data source or reviewing the data extraction and loading logs.
   - **Impact Assessment:** Evaluate how critical these columns are to your analysis or models. For instance, if `VendorID` or `payment_type` plays a significant role in your analysis, more sophisticated methods of imputation might be necessary.

2. **Deletion:**
   - **Remove Rows:** If the missing data represents a small proportion of the total dataset and is not systematically biased, you might consider dropping these rows.
   - **Remove Columns:** If a column has a significant number of missing values that cannot be imputed accurately and is not critical, consider dropping the column.

3. **Imputation:**
   - **Categorical Imputation:** For categorical fields like `VendorID`, `RatecodeID`, `store_and_fwd_flag`, and `payment_type`, you can impute missing values using the mode (the most frequently occurring value in each column) or employ predictive imputation methods (e.g., classification algorithms if patterns can be discerned from other data).
   - **Count Imputation:** For `passenger_count`, consider using median values for imputation, as mean might be skewed by outliers. Alternatively, use a model-based approach if the missing data pattern suggests dependency on other variables (e.g., trip distance, fare amount).

4. **Advanced Techniques:**
   - **Predictive Modeling:** Use machine learning models to predict missing values based on other variables in the dataset.
   - **Multiple Imputation:** Multiple imputation techniques can be used for handling missing data in a way that reflects the uncertainty about the right approach to impute.


In [7]:
print("Data types of each column:\n", df1.dtypes)


Data types of each column:
 VendorID                       category
tpep_pickup_datetime     datetime64[ns]
tpep_dropoff_datetime    datetime64[ns]
passenger_count                 float64
trip_distance                   float64
RatecodeID                     category
store_and_fwd_flag             category
PULocationID                   category
DOLocationID                   category
payment_type                   category
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
improvement_surcharge           float64
total_amount                    float64
congestion_surcharge            float64
dtype: object


### Partition 2 Initial Exploration

In [8]:
# Display the first few rows of the first DataFrame
print(df2.head(3))

# Display summary information about DataFrame
print(df2.info())

# Basic statistical details
print(df2.describe())


  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  passenger_count  trip_distance RatecodeID store_and_fwd_flag PULocationID DOLocationID payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  improvement_surcharge  total_amount  congestion_surcharge
0      2.0  05/15/2021 10:35:13 AM  05/15/2021 11:08:19 AM             5.00          17.89        2.0                  N        132.0        224.0          1.0        52.00   0.00     0.50        8.00          6.55                   0.30         69.85                  2.50
1      2.0  05/15/2021 10:13:57 AM  05/15/2021 10:19:57 AM             2.00           1.11        1.0                  N        114.0        249.0          1.0         6.50   0.00     0.50        1.96          0.00                   0.30         11.76                  2.50
2      2.0  05/15/2021 10:26:33 AM  05/15/2021 10:43:56 AM             2.00           4.44        1.0                  N        100.0         12.0          1.0        16.50   0.0

In [9]:
# Count missing values in each column
missing_values2 = df2.isnull().sum()
print("Missing values in each column:\n", missing_values2)


Missing values in each column:
 VendorID                 381490
tpep_pickup_datetime          0
tpep_dropoff_datetime         0
passenger_count          381490
trip_distance                 0
RatecodeID               381490
store_and_fwd_flag       381490
PULocationID                  0
DOLocationID                  0
payment_type             381490
fare_amount                   0
extra                         0
mta_tax                       0
tip_amount                    0
tolls_amount                  0
improvement_surcharge         0
total_amount                  0
congestion_surcharge          0
dtype: int64


In [13]:
print("Data types of each column:\n", df2.dtypes)


Data types of each column:
 VendorID                 category
tpep_pickup_datetime       object
tpep_dropoff_datetime      object
passenger_count           float64
trip_distance             float64
RatecodeID               category
store_and_fwd_flag       category
PULocationID             category
DOLocationID             category
payment_type             category
fare_amount               float64
extra                     float64
mta_tax                   float64
tip_amount                float64
tolls_amount              float64
improvement_surcharge     float64
total_amount              float64
congestion_surcharge      float64
dtype: object


### Partition 2 Initial Exploration

In [11]:
# Display the first few rows of the first DataFrame
print(df3.head(3))

# Display summary information about DataFrame
print(df3.info())

# Basic statistical details
print(df3.describe())


  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  passenger_count  trip_distance RatecodeID store_and_fwd_flag PULocationID DOLocationID payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  improvement_surcharge  total_amount  congestion_surcharge
0      1.0  08/10/2021 04:22:40 PM  08/10/2021 04:40:02 PM             1.00           2.50        1.0                  N        170.0        236.0          1.0        13.00   3.50     0.50        3.45          0.00                   0.30         20.75                  2.50
1      1.0  08/10/2021 04:11:09 PM  08/10/2021 04:19:01 PM             0.00           1.00        1.0                  N        107.0         90.0          2.0         7.00   3.50     0.50        0.00          0.00                   0.30         11.30                  2.50
2      1.0  08/10/2021 04:20:56 PM  08/10/2021 04:35:54 PM             0.00           2.40        1.0                  N         68.0        142.0          1.0        11.50   3.5

In [17]:
# Count missing values in each column
missing_values3 = df3.isnull().sum()
print("Missing values in each column:\n", missing_values3)

Missing values in each column:
 VendorID                      0
tpep_pickup_datetime          0
tpep_dropoff_datetime         0
passenger_count          278444
trip_distance                 0
RatecodeID               278444
store_and_fwd_flag       278444
PULocationID                  0
DOLocationID                  0
payment_type                  0
fare_amount                   0
extra                         0
mta_tax                       0
tip_amount                    0
tolls_amount                  0
improvement_surcharge         0
total_amount                  0
congestion_surcharge     278444
dtype: int64


In [15]:
print("Data types of each column:\n", df3.dtypes)

Data types of each column:
 VendorID                 category
tpep_pickup_datetime       object
tpep_dropoff_datetime      object
passenger_count           float64
trip_distance             float64
RatecodeID               category
store_and_fwd_flag       category
PULocationID             category
DOLocationID             category
payment_type             category
fare_amount               float64
extra                     float64
mta_tax                   float64
tip_amount                float64
tolls_amount              float64
improvement_surcharge     float64
total_amount              float64
congestion_surcharge      float64
dtype: object


### Partition 4 Initial Exploration

In [16]:
# Display the first few rows of the first DataFrame
print(df4.head(3))

# Display summary information about DataFrame
print(df4.info())

# Basic statistical details
print(df4.describe())

  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  passenger_count  trip_distance RatecodeID store_and_fwd_flag PULocationID DOLocationID payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  improvement_surcharge  total_amount  congestion_surcharge
0      2.0  10/28/2021 03:35:00 PM  10/28/2021 03:44:01 PM             1.00           1.80        1.0                  N        170.0         79.0          1.0         8.00   0.00     0.50        1.00          0.00                   0.30         12.30                  2.50
1      2.0  10/28/2021 03:45:48 PM  10/28/2021 04:04:51 PM             1.00           1.45        1.0                  N         79.0        170.0          1.0        12.50   0.00     0.50        3.16          0.00                   0.30         18.96                  2.50
2      1.0  10/28/2021 03:08:00 PM  10/28/2021 03:26:20 PM             2.00           1.70        1.0                  N        141.0        142.0          1.0        12.00   2.5

In [18]:
# Count missing values in each column
missing_values4 = df4.isnull().sum()
print("Missing values in each column:\n", missing_values4)

Missing values in each column:
 VendorID                      0
tpep_pickup_datetime          0
tpep_dropoff_datetime         0
passenger_count          366223
trip_distance                 0
RatecodeID               366223
store_and_fwd_flag       366223
PULocationID                  0
DOLocationID                  0
payment_type                  0
fare_amount                   0
extra                         0
mta_tax                       0
tip_amount                    0
tolls_amount                  0
improvement_surcharge         0
total_amount                  0
congestion_surcharge     366223
dtype: int64


In [19]:
print("Data types of each column:\n", df4.dtypes)

Data types of each column:
 VendorID                 category
tpep_pickup_datetime       object
tpep_dropoff_datetime      object
passenger_count           float64
trip_distance             float64
RatecodeID               category
store_and_fwd_flag       category
PULocationID             category
DOLocationID             category
payment_type             category
fare_amount               float64
extra                     float64
mta_tax                   float64
tip_amount                float64
tolls_amount              float64
improvement_surcharge     float64
total_amount              float64
congestion_surcharge      float64
dtype: object


## Summary of initial Exploration and Next Steps:

